pipeline_dp/pipeline_operations.py

```diff
@@ -349,3 +346,213 @@ def count_per_element(self, col, stage_name: typing.Optional[str] = None):
+    yield from iter_obj
+    yield Sentinel.EOI
+
+
+def _multiproc_iter_output(iter_obj):
+    for item in iter_obj:
+        if item is Sentinel.EOI:
+            break
+        yield item
+
+
+# workaround for passing lambda functions to multiprocessing
+# according to https://medium.com/@yasufumy/python-multiprocessing-c6d54107dd55
+_pool_current_func = None
+
+
+def _pool_worker_init(func):
+    global _pool_current_func
+    _pool_current_func = func
+
+
+def _pool_worker(row):
+    if row is Sentinel.EOI:
+        return Sentinel.EOI
+    return _pool_current_func(row)
+
+
+class _LazyMultiProcIterator:
+    def __init__(
+        self,
+        job: typing.Callable,
+        job_inputs: typing.Iterable,
+        chunksize: int = 1,
+        n_jobs: typing.Optional[int] = None,
+        **pool_kwargs
+    ):
+        self.job = job
+        self.chunksize = chunksize
+        self.job_inputs = job_inputs
+        self.n_jobs = n_jobs
+        self.pool_kwargs = pool_kwargs
+        self._outputs = None  # type: typing.Optional[typing.Iterator]
+        self._pool = None
+
+    def _init_pool(self):
+        self._pool = mp.Pool(
+            self.n_jobs,
+            initializer=_pool_worker_init,
+            initargs=(self.job,),
+            **self.pool_kwargs
+        )
+        return self._pool
+
+    def _trigger_iterations(self):
+        if self._outputs is None:
+            self._outputs = self._init_pool().map(
+                _pool_worker, _multiproc_iter_input(self.job_inputs), self.chunksize
+            )
+
+    def __iter__(self):
+        if isinstance(self.job_inputs, _LazyMultiProcIterator):
+            self.job_inputs._trigger_iterations()
+        self._trigger_iterations()
+        yield from _multiproc_iter_output(self._outputs)
+
+
+class _LazyMultiProcGroupByIterator(_LazyMultiProcIterator):
+    def __init__(self, job_inputs: typing.Iterable,
+                chunksize: int,
+                n_jobs: typing.Optional[int],
+                **pool_kwargs):
+        self.manager = mp.Manager()
+        self.results_dict = self.manager.dict()
+        def insert_row(captures, row):
+            (results_dict_,) = captures
+            key, val = row
+            results_dict_[key].append(val)
+
+        insert_row = partial(insert_row, (self.results_dict,))
+
+        super().__init__(insert_row, job_inputs, 
+                         chunksize=chunksize, 
+                         n_jobs=n_jobs, 
+                         **pool_kwargs)
+
+    def _trigger_iterations(self):
+        if self._outputs is None:
+            keys = set(k for k, v in self.job_inputs)
+            self.results_dict.update({k: self.manager.list() for k in keys})
+            self._init_pool().map(_pool_worker, self.job_inputs, self.chunksize)
+            self._outputs = ((k, list(v)) for k, v in self.results_dict.items())
+
+class _LazyMultiProcCountIterator(_LazyMultiProcIterator):
+    def __init__(self, job_inputs: typing.Iterable,
```

Comment:
Could you please add documentation for this class with a short explanation how it works? With mentioning that results_dict is used for collecting results between processes and that it protects us from any parallel issues.

Reply:
Done, hope to satisfaction.

Reply:
Thanks!